### PR TITLE
#1535 Two Way Message Enhancement: Support for All Clients

### PR DESCRIPTION
**Pull request by EvanParish**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

This PR is for adding support for alternative clients, in addition to vetext for two-way sms messages.

issue #1535 

## Type of change

Please check the relevant option(s).

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Documentation changes only

## How Has This Been Tested?

How has this been tested? (e.g. Unit tests, Tested locally, Tested as a GitHub action, Depoyed to dev, etc)  
Please provide relevant information and / or links to demonstrate the functionality or fix. (e.g. screenshots, link to deployment, regression test results, etc)

- [x] Unit tests pass
- [x] Deployed to Staging
- [x] Two-way testing complete, logs checked and screenshots provided

Vetext successful connection:
image

Vetext successful connection (staging.api.vetext):
image

PeriOptimization successful connection:
image

## Checklist

- [x] I have assigned myself to this PR
- [x] PR has an appropriate title: #9999 - What the thing does
- [ ] PR has a detailed description, including links to specific documentation
- [x] I have added the [appropriate labels](https://github.com/department-of-veterans-affairs/notification-api/blob/master/.github/release.yaml) to the PR.
- [x] I did not remove any parts of the template, such as checkboxes even if they are not used
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to any documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have ensured the latest master is merged into my branch and all checks are green prior to review
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] The ticket is now moved into the DEV test column

### CODE REVIEW

File: documents/openapi/openapi.yaml [hunk 1]
```diff
@@ -2510,9 +2570,64 @@ components:
...
+          format: uuid
+          nullable: true
+        url_endpoint:
+          type: string
     InboundNumberCreateResponse:
       type: object
-      properties: {} # TODO (issue 965)
+      properties:
+        data:
+          $ref: "#components/schemas/InboundNumberResponseData"
```
Comment (cris-oddball) [inline documents/openapi/openapi.yaml:2623]: This line is causing three validation errors. 
Please update as shown, with a `/ `after `#` and before `components`
`$ref: "#/components/schemas/InboundNumberResponseData"`

**Review by cris-oddball — CHANGES_REQUESTED** [review]

File: documents/openapi/openapi.yaml [hunk 2]
```diff
@@ -1467,6 +1471,13 @@ paths:
           $ref: "#/components/responses/ServiceUnavailable"
     post:
       summary: Create an inbound number
+      tags:
+        - inbound-number
+      requestBody:
```
Comment (cris-oddball) [inline documents/openapi/openapi.yaml:1476]: this needs `required: true`

File: documents/openapi/openapi.yaml [hunk 3]
```diff
@@ -1467,6 +1471,13 @@ paths:
...
+        content:
+          application/json:
+            schema: 
+              $ref: "#/components/schemas/InboundNumberRequestData"
```
Comment (cris-oddball) [inline documents/openapi/openapi.yaml:1480]: this needs to have required fields as well

File: documents/openapi/openapi.yaml [hunk 4]
```diff
@@ -1484,42 +1495,91 @@ paths:
...
+        "500":
+          $ref: "#/components/responses/InternalServerError"
+        "502":
+          $ref: "#/components/responses/BadGateway"
+        "503":
+          $ref: "#/components/responses/ServiceUnavailable"
   /inbound-number/{inbound_number_id}:
     post:
       summary: Update an inbound number
+      tags:
```
Comment (cris-oddball) [inline documents/openapi/openapi.yaml:1519]: this needs `required: true`

**Review by cris-oddball — CHANGES_REQUESTED** [review]
Please also update the swagger version on line 4 from `version: 1.0.4` to `version: 1.0.5`

File: documents/openapi/openapi.yaml (hunk 4 above)
Comment (EvanParish) [reply documents/openapi/openapi.yaml:1538]: added

File: documents/openapi/openapi.yaml (hunk 3 above)
Comment (EvanParish) [reply documents/openapi/openapi.yaml]: added

File: documents/openapi/openapi.yaml (hunk 2 above)
Comment (EvanParish) [reply documents/openapi/openapi.yaml:1476]: done

File: documents/openapi/openapi.yaml (hunk 1 above)
Comment (EvanParish) [reply documents/openapi/openapi.yaml]: updated

File: lambda_functions/two_way_sms/two_way_sms_v2.py [hunk 5]
```diff
@@ -271,25 +264,40 @@ def valid_event(event_data: dict) -> bool:
...
-        'Authorization': 'Basic ' + vetext_auth_token
+        'Authorization': 'Basic ' + client_auth_token
     }
 
     try:
         response = requests.post(
-            url,            
-            verify=False,
+            url,
+            verify=False if auth_parameter and 'vetext' in auth_parameter else True,
```
Comment (k-macmillan) [inline lambda_functions/two_way_sms/two_way_sms_v2.py:300]: I don't believe we can make it down here with `auth_parameter` being None or an empty string.

File: tests/app/db.py [hunk 6]
```diff
@@ -591,7 +591,8 @@ def create_inbound_number(
     active=True,
     service_id=None,
     url_endpoint=None,
-    self_managed=False
+    self_managed=False,
+    auth_parameter=None
```
Comment (k-macmillan) [inline tests/app/db.py:595]:
```suggestion
    auth_parameter=None,
```

File: tests/app/db.py [hunk 7]
```diff
@@ -600,7 +601,8 @@ def create_inbound_number(
         active=active,
         service_id=service_id,
         url_endpoint=url_endpoint,
-        self_managed=self_managed
+        self_managed=self_managed,
+        auth_parameter=auth_parameter
```
Comment (k-macmillan) [inline tests/app/db.py:605]:
```suggestion
        auth_parameter=auth_parameter,
```

File: tests/lambda_functions/two_way_sms/test_two_way_sms_v2.py [hunk 8]
```diff
@@ -40,124 +62,179 @@ def test_validate_event_body(mocker, event):
...
+    assert response is False
+
 
 def test_forward_to_service_failed_on_general_exception(mocker):
     mocker.patch(f'{LAMBDA_MODULE}.requests.post', side_effect=Exception)
 
     with pytest.raises(Exception):
         result = forward_to_service({}, 'https://someurl.com')
 
+        print(f'evan: {result}')
```
Comment (k-macmillan) [inline tests/lambda_functions/two_way_sms/test_two_way_sms_v2.py:126]:
```suggestion
```

**Review by k-macmillan — CHANGES_REQUESTED** [review]

File: lambda_functions/two_way_sms/two_way_sms_v2.py [hunk 9]
```diff
@@ -271,25 +264,40 @@ def valid_event(event_data: dict) -> bool:
...
+            client_auth_token = get_ssm_param_info(client_api_auth_ssm_path=auth_parameter)
+        except Exception as e:
+            logger.exception("Issue attempting to get ssm parameter for incoming two-way sms - Exception: %s", e)
+
+        if client_auth_token is None:
+            logger.critical(
+                'Raising execption because client_auth_token could not be retrieved from SSM. client_auth_token = %s',
+                client_auth_token
+            )
+            raise Exception('Raising execption because client_auth_token could not be retrieved from SSM.')
```
Comment (cris-oddball) [inline lambda_functions/two_way_sms/two_way_sms_v2.py:285]: typo in word `exception` here

**Review by cris-oddball — CHANGES_REQUESTED** [review]

File: lambda_functions/two_way_sms/two_way_sms_v2.py (hunk 5 above)
Comment (EvanParish) [reply lambda_functions/two_way_sms/two_way_sms_v2.py:300]: I was hitting an issue there during testing. Maybe the tests just need reworked a little bit.

File: lambda_functions/two_way_sms/two_way_sms_v2.py (hunk 9 above)
Comment (EvanParish) [reply lambda_functions/two_way_sms/two_way_sms_v2.py]: fixed

File: tests/lambda_functions/two_way_sms/test_two_way_sms_v2.py (hunk 8 above)
Comment (EvanParish) [reply tests/lambda_functions/two_way_sms/test_two_way_sms_v2.py]: removed

File: tests/app/db.py (hunk 7 above)
Comment (EvanParish) [reply tests/app/db.py]: added

File: tests/app/db.py (hunk 6 above)
Comment (EvanParish) [reply tests/app/db.py]: added

File: lambda_functions/two_way_sms/two_way_sms_v2.py [hunk 10]
```diff
@@ -182,12 +185,13 @@ def notify_incoming_sms_handler(event: dict, context: any):
         if not valid_message_body(inbound_sms):
             logger.error("The record's message body is invalid: %s", inbound_sms)
 
-            # Push the record to the dead letter queue, and return 200 to have the message removed from the feeder queue.
+            # Push the record to the dead letter queue, and return 200 to have the message removed from the feeder queue
             push_to_sqs(record_body, False)
             if len(event["Records"]) > 1:
                 # TODO - Remove this block as part of #1014.  The problem is that the handler returns if any record has
                 # an invalid message.  It's not a problem now because we only expect one record.
```
Comment (cris-oddball) [inline lambda_functions/two_way_sms/two_way_sms_v2.py:192]: We can remove this TODO now since we are going to close that ticket. Check with Kyle first!

**Review by cris-oddball — CHANGES_REQUESTED** [review]
Kyle, Melanie and I talked about about removing the TODO mentioned in this review earlier today. Please check with Kyle.

File: lambda_functions/two_way_sms/two_way_sms_v2.py (hunk 5 above)
Comment (EvanParish) [reply lambda_functions/two_way_sms/two_way_sms_v2.py]: adjusted unit tests to send a str instead of `None` type, and removed extra `auth_parameter` check

**Review by k-macmillan — APPROVED** [review]

**Review by cris-oddball — APPROVED** [review]

**Pull request merged**